src/inter_dc_log_sender.erl

```diff
@@ -0,0 +1,138 @@
+  encode_handoff_item/2,
+  is_empty/1,
+  terminate/2,
+  delete/1]).
+
+-record(state, {
+  partition :: partition_id(),
+  buffer, %% log_tx_assembler:state
+  last_log_id :: log_opid(),
+  ping_timer :: any()
+}).
+
+%% API
+start_vnode(I) -> riak_core_vnode_master:get_vnode_pid(I, ?MODULE).
+send(Partition, Operation) -> dc_utilities:call_vnode_sync(Partition, inter_dc_log_sender_master, {log_event, Operation}).
+ping(Partition) -> dc_utilities:call_vnode_sync(Partition, inter_dc_log_sender_master, ping).
+
+init([Partition]) ->
+  {ok, set_timer(#state{
+    partition = Partition,
+    buffer = log_txn_assembler:new_state(),
+    last_log_id = 0,
+    ping_timer = none
+  })}.
+
+handle_command(ping, _Sender, State) ->
+  %% TODO: think if the timestamp could cause any problems
+  PingTxn = inter_dc_txn:ping(State#state.partition, State#state.last_log_id, get_stable_time(State#state.partition)),
+  {reply, ok, broadcast(State, PingTxn)};
+
+handle_command({log_event, Operation}, _Sender, State) ->
+  {Result, NewBufState} = log_txn_assembler:process(Operation, State#state.buffer),
+  State1 = State#state{buffer = NewBufState},
+  State2 = case Result of
+    {ok, Ops} ->
+      Txn = inter_dc_txn:from_ops(Ops, State1#state.partition, State#state.last_log_id),
+      %% sanity check - we only publish locally committed transactions
+      case inter_dc_txn:is_local(Txn) of
+        true -> broadcast(State1, Txn);
+        false -> State1
+      end;
+    none -> State1
+  end,
+  {reply, ok, State2}.
+
+
+handle_coverage(_Req, _KeySpaces, _Sender, State) -> {stop, not_implemented, State}.
+handle_exit(_Pid, _Reason, State) -> {noreply, State}.
+handoff_starting(_TargetNode, State) -> {true, State}.
+handoff_cancelled(State) -> {ok, State}.
+handoff_finished(_TargetNode, State) -> {ok, State}.
+handle_handoff_command( _Message , _Sender, State) -> {noreply, State}.
+handle_handoff_data(_Data, State) -> {reply, ok, State}.
+encode_handoff_item(Key, Operation) -> term_to_binary({Key, Operation}).
+is_empty(State) -> {true, State}.
+terminate(_Reason, _State) -> ok.
+delete(State) -> {ok, State}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%
+
+set_timer(State) ->
+  ClearedState = clr_timer(State),
+  {ok, Timer} = timer:apply_after(1000, inter_dc_log_sender, ping, [State#state.partition]),
+  ClearedState#state{ping_timer = Timer}.
+
+clr_timer(State = #state{ping_timer = none}) -> State;
+clr_timer(State = #state{ping_timer = Timer}) ->
+  {ok, cancel} = timer:cancel(Timer),
+  State#state{ping_timer = none}.
+
+broadcast(State, Txn) ->
+  State1 = clr_timer(State),
+  inter_dc_pub:broadcast(Txn),
+  Id = inter_dc_txn:last_log_opid(Txn),
+  set_timer(State1#state{last_log_id = Id}).
+
+%% @doc Return smallest snapshot time of active transactions.
+%%      No new updates with smaller timestamp will occur in future.
+get_stable_time(_Partition) ->
+    Now = inter_dc_utils:now_millisec(),
+    Now.
+%% case clocksi_vnode:get_active_txns_call(Partition) of
```

Comment:
This was commented out because I saw it was a bottleneck, so it is incorrect, I have a fix that shouldn't be a bottleneck that I will make tomorrow.